server/lib/messaging/process-chat-message.ts

```diff
@@ -0,0 +1,36 @@
+import { matchUserMentions, MENTION_REGEX } from '../../../common/text/mentions'
+import { SbUser } from '../../../common/users/user-info'
+import { findUsersByName } from '../users/user-model'
+
+/**
+ * Processes the chat message by extracting the user mentions in it and replacing them with a markup
+ * that includes the user ID for each mention. The markup is in the format of `<@USER_ID>`, which
+ * can then easily be used on the client to replace the markup with the appropriate UI.
+ *
+ * Any mention that doesn't map to a valid user in our system is left unreplaced.
+ *
+ * @returns A tuple of the processed chat message, along with a map of userId -> user of each
+ *   mentioned user.
+ */
+export async function processMessageContents(
+  text: string,
+  allowedMentionUsers: Set<string>,
```

Comment:
I haven't reviewed the rest of this yet, but I'll mention this now since it's something you can change earlier:

I don't think this is desirable or good. I think mentioning people that aren't present is fine, it makes the username clickable, you can use it to whisper them, access their profile, etc. Alongside that, doing this kind of filtering can be VERY expensive.

With this change, you've made it so we have to look up all the users in #ShieldBattery on *every message* to #ShieldBattery, so we're grabbing and returning thousands of users every single time, converting all their names to lowercase, putting them in a Set... That's gonna wreak havoc on stuff and make everything slower.

Just leave this how it was before, there's really no harm in mentioning a user that isn't present. It's a different story when we're dealing with what to auto-complete and what triggers a notification, but this code here really has nothing to do with either of those things.